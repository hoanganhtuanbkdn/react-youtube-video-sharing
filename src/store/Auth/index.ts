export * from './Auth.redux';
export * from './Auth.saga';
