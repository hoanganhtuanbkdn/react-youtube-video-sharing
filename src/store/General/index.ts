export * from './General.redux';
export * from './General.saga';
