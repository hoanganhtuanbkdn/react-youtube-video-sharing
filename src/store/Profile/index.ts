export * from './Profile.redux';
export * from './Profile.saga';
