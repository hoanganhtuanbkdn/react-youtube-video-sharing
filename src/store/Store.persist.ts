import createWebStorage from 'redux-persist/lib/storage/createWebStorage';

// https://github.com/vercel/next.js/discussions/15687
const createNoopStorage = () => {
	return {
		getItem(_key: string) {
			return Promise.resolve(null);
		},
		setItem(_key: string, value: any) {
			return Promise.resolve(value);
		},
		removeItem(_key: string) {
			return Promise.resolve();
		},
	};
};

const storage =
	typeof window !== 'undefined'
		? createWebStorage('local')
		: createNoopStorage();

export const persistConfig = {
	key: 'root',
	version: 1,
	storage,
	whitelist: ['general'],
};
