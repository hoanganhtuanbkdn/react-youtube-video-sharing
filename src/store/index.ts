import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

import { store } from './Store.redux';

export * from './Auth';
export * from './Profile';
export * from './Store.redux';

export type RootState = ReturnType<typeof store.getState>;
export const GlobalDispatch = store.dispatch;

export type AppDispatch = typeof GlobalDispatch;
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useAppDispatch: () => AppDispatch = useDispatch;
