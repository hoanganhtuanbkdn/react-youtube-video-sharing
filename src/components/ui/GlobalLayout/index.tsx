import { socket } from '@/services/socket';
import { GlobalDispatch, ProfileActions, ProfileSelectors } from '@/store';
import React, { ReactNode, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import Header from '../Header';

import { notification } from 'antd';
import { useRouter } from 'next/router';
import { split } from 'ramda';

const GlobalLayout: React.FC<{ children: ReactNode }> = ({ children }) => {
	const [isConnected, setIsConnected] = useState(socket.connected);

	const profile = useSelector(ProfileSelectors.selectProfile);

	const router = useRouter();

	const onNewMessages = (value: any) => {
		if (!profile.email) return;
		try {
			const raw = split('{', value);

			const message = JSON.parse(`{${raw?.[1]}`);

			if (profile.email === message?.email) return;

			notification.open({
				message: `${message?.email} shared a movie`,
				description: message?.title,
				onClick: () => {
					router.replace('/');
				},
			});
		} catch (e) {}
	};

	useEffect(() => {
		GlobalDispatch(ProfileActions.getProfileRequest());

		function onConnect() {
			setIsConnected(true);
		}

		function onDisconnect() {
			setIsConnected(false);
		}

		socket.on('connect', onConnect);
		socket.on('disconnect', onDisconnect);
		socket.on('some room event', function (msg) {
			// console.log('msg', msg);
		});

		return () => {
			socket.off('connect', onConnect);
			socket.off('disconnect', onDisconnect);
			socket.off('chat message');
			socket.off('some room event');
		};
	}, []);

	useEffect(() => {
		socket.on('chat message', onNewMessages);
	}, [profile]);

	return (
		<main className={'col scrollbar-md min-h-screen w-screen'}>
			<Header isConnected={isConnected} />
			<div className="container min-h-full">{children}</div>
		</main>
	);
};

export default GlobalLayout;
