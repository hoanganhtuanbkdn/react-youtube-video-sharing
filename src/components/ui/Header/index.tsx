import SmartLink from '@/components/SmartLink';
import { ROUTERS } from '@/constant';
import { AuthActions, GlobalDispatch, ProfileSelectors } from '@/store';
import React from 'react';
import { useSelector } from 'react-redux';

import AuthModal from '../AuthModal';

import { Button } from 'antd';
import { useRouter } from 'next/router';

export default function Header() {
	const profile = useSelector(ProfileSelectors.selectProfile);
	const router = useRouter();
	const onShowModal = (isRegister: boolean) => {
		GlobalDispatch(
			AuthActions.setShowAuthModal({
				status: true,
				type: isRegister ? 'REGISTER' : 'LOGIN',
			})
		);
	};

	const onLogout = () => {
		GlobalDispatch(AuthActions.logoutRequest(router));
	};

	return (
		<div className=" sticky top-0 z-40 h-[90px] bg-white shadow">
			<div className=" container ">
				<div className="row-between px-3 md:px-0">
					<h1>Funny Movie</h1>
					{profile?.id ? (
						<div className="col md:row text-xs md:gap-4 md:text-base">
							<p>Welcome {[profile?.email]}</p>
							<div className="row nd:gap-4 gap-2">
								<SmartLink href={ROUTERS.SHARE}>
									<Button type="primary">
										Share a movie
									</Button>
								</SmartLink>
								<Button onClick={onLogout}>Logout</Button>
							</div>
						</div>
					) : (
						<div className="row nd:gap-4 gap-2">
							<Button
								type="primary"
								onClick={() => onShowModal(false)}
							>
								Login
							</Button>
							<Button onClick={() => onShowModal(true)}>
								Register
							</Button>
							<AuthModal />
						</div>
					)}
				</div>
			</div>
		</div>
	);
}
