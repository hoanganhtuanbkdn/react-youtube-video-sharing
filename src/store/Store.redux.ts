import { configureStore } from '@reduxjs/toolkit';

import { AuthReducer } from './Auth/Auth.redux';
import { GeneralReducer } from './General';
import { ProfileReducer } from './Profile';
import { persistConfig } from './Store.persist';
import rootSaga from './Store.saga';

import { combineReducers } from 'redux';
import {
	FLUSH,
	PAUSE,
	PERSIST,
	persistReducer,
	persistStore,
	PURGE,
	REGISTER,
	REHYDRATE,
} from 'redux-persist';
import createSagaMiddleware from 'redux-saga';

const rootReducer = combineReducers({
	auth: AuthReducer,
	profile: ProfileReducer,
	general: GeneralReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);
const sagaMiddleware = createSagaMiddleware();
const store = configureStore({
	reducer: persistedReducer,
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware({
			serializableCheck: {
				ignoredActions: [
					FLUSH,
					REHYDRATE,
					PAUSE,
					PERSIST,
					PURGE,
					REGISTER,
				],
			},
		}).concat(sagaMiddleware),
});

const persistor = persistStore(store);

sagaMiddleware.run(rootSaga);

export { persistor, store };
