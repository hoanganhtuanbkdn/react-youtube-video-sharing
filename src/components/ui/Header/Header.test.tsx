import { store } from '@/store';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';

import GlobalLayout from '../GlobalLayout';
import Header from '.';

import '@testing-library/jest-dom/extend-expect';
import mockRouter from 'next-router-mock';

jest.mock('next/router', () => require('next-router-mock'));

Object.defineProperty(window, 'matchMedia', {
	writable: true,
	value: jest.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: jest.fn(), // deprecated
		removeListener: jest.fn(), // deprecated
		addEventListener: jest.fn(),
		removeEventListener: jest.fn(),
		dispatchEvent: jest.fn(),
	})),
});
describe('next-router-mock', () => {
	test('renders Header correctly', () => {
		const { container } = render(
			<Provider store={store}>
				<Header isConnected isProfile />
			</Provider>
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	test('check login button', () => {
		render(
			<Provider store={store}>
				<Header isConnected />
			</Provider>
		);

		const button = screen.getByTestId('loginBtn');
		fireEvent.click(button);

		expect(screen.getByText('Funny Login')).toBeInTheDocument();
	});
	test('check register button', () => {
		render(
			<Provider store={store}>
				<Header isConnected />
			</Provider>
		);

		const button = screen.getByTestId('registerBtn');
		fireEvent.click(button);

		expect(screen.getByText('Funny Register')).toBeInTheDocument();
	});
	test('check share button', () => {
		// Set the initial url:
		mockRouter.push('/share');

		// Render the component:
		render(
			<Provider store={store}>
				<Header isConnected isProfile />
			</Provider>
		);

		const button = screen.getByTestId('shareBtn');

		// Click the button:
		fireEvent.click(button);

		// Ensure the router was updated:
		expect(mockRouter).toMatchObject({
			asPath: '/share',
			pathname: '/share',
		});
	});

	test('check logout button', () => {
		render(
			<Provider store={store}>
				<GlobalLayout>
					<Header isConnected isProfile />
				</GlobalLayout>
			</Provider>
		);

		const button = screen.getByTestId('logoutBtn');
		fireEvent.click(button);

		expect(screen.getByText('Login')).toBeInTheDocument();
	});
});
