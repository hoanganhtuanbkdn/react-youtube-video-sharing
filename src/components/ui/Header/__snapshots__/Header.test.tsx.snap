// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`next-router-mock renders Header correctly 1`] = `
<div
  class=" sticky top-0 z-40 h-[90px] bg-white shadow"
>
  <div
    class="container "
  >
    <div
      class="row-between px-3 md:px-0"
    >
      <a
        href="/"
      >
        <h1
          class="logo"
        >
          Funny Movie
        </h1>
      </a>
      <div
        class="col md:row text-xs md:gap-4 md:text-base"
      >
        <p>
          Welcome 
        </p>
        <div
          class="row nd:gap-4 gap-2"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-w8mnev ant-btn-primary"
            data-testid="shareBtn"
            type="button"
          >
            <span>
              Share a movie
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-w8mnev ant-btn-default"
            data-testid="logoutBtn"
            type="button"
          >
            <span>
              Logout
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
